airbyte-integrations/connectors/source-outreach/source_outreach/source.py

```diff
@@ -121,34 +188,268 @@ def path(self, **kwargs) -> str:
         return "sequenceStates"
 
 
+# CreatedAt incremental stream
+class CreatedAtIncrementalOutreachStream(OutreachStream, ABC):
+    @property
+    def cursor_field(self) -> str:
+        return "createdAt"
+
+    def get_updated_state(
+        self,
+        current_stream_state: MutableMapping[str, Any],
+        latest_record: Mapping[str, Any],
+    ) -> Mapping[str, Any]:
+        current_stream_state = current_stream_state or {}
+
+        current_stream_state_date = current_stream_state.get(
+            "createdAt", self.start_date
+        )
+        latest_record_date = latest_record.get("attributes", {}).get(
+            "createdAt", self.start_date
+        )
+
+        return {"createdAt": max(current_stream_state_date, latest_record_date)}
+
+    def request_params(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> MutableMapping[str, Any]:
+        params = super().request_params(
+            stream_state=stream_state,
+            stream_slice=stream_slice,
+            next_page_token=next_page_token,
+        )
+        params["sort"] = "createdAt"
+        if "createdAt" in stream_state:
+            params["filter[createdAt]"] = stream_state["createdAt"] + "..inf"
+        return params
+
+
+class Events(CreatedAtIncrementalOutreachStream):
+    """
+    Event stream. Yields data from the GET /events endpoint.
+    See https://api.outreach.io/api/v2/docs#events
+    """
+
+    primary_key = "id"
+
+    def path(self, **kwargs) -> str:
+        return "events"
+
+
+# Dependent stream
+class DependentOutreachStream(OutreachStream, ABC):
+    def __init__(
+        self,
+        **kwargs,
+    ):
+        self.dependent_api_response_json = None
+        self.remaining_ids = None
+        super().__init__(**kwargs)
+
+    def _fill_out_remaining_ids(self) -> None:
```

Comment:
Can you add some comments to this function?  Hard for me to understand what's going on.

Reply:
Added documentation, but yeah this area is a lot of haxing to work around a bug in Outreach's code, and not designed for how airbyte like's to consume streams.